principles.rst

```diff
@@ -303,3 +309,130 @@ down GHC even when those features are not active. Yet this principle is importan
 as we hope not to make the current situation worse.
 
 From `#378`_, slightly generalized.
+
+
+
+GHC stability principles
+--------------------------
+
+The Haskell ecosystem has a built-in tension between stability and innovation.
+
+One the one hand, breaking changes impose heavy costs:
+
+* Users keep having to update their code
+* Library authors keep having to update their code
+* Updating to a later version of a library can in turn force taking on new dependencies.
+
+These difficulties add friction, discourage adoption, and make Haskell an unpredictable investment.
+
+On the other hand, GHC has always sought to be a laboratory for innovation.  If the stability guarantees are too onerous, we risk imposing a different set of costs:
+
+* Volunteers may get discouraged by the hoops they have to jump through to get a change agreed.
+* The language may become stuck in a local optimum, because moving to a better design would require breaking changes.
+* GHC may become encrusted with cruft and workarounds that exist to preserve existing, but undesirable, behaviour.
+
+We can't navigate this tension with simple all-or-nothing rules.  We have to take each case on its merits, aware of both sets of costs.  Rather than starting the debate from scratch each time, we can have some general rules, or principles, to guide us.  The `GHC base library proposal <https://github.com/haskellfoundation/tech-proposals/blob/main/proposals/accepted/051-ghc-base-libraries.rst>`_ made a start in this direction; this section discusses GHC itself.
+
+Assumptions
+~~~~~~~~~~~~
+
+We assume:
+
+* That we have adopted the  `GHC base library proposal <https://github.com/haskellfoundation/tech-proposals/blob/main/proposals/accepted/051-ghc-base-libraries.rst>`_, which establishes the ``ghc-experimental`` package.
+* That we have identified some extensions as "Experimental".  These extensions may well change, and are subject to
+  much weaker stability goals.  `GHC Proposal #601 <https://github.com/ghc-proposals/ghc-proposals/pull/601>`_, describes a plan along these lines, but all we need for this document is a list of Experimental extensions.
+
+Terminology
+~~~~~~~~~~~~~~~
+
+We define a **stable Haskell package** as follows. A stable Haskell package
+
+* Does not use Experimental extensions.
+* Does not use experimental features.  (Examples: use of the Javascript or Wasm back end, builds on non-tier-1 platforms.)
+* Does not rely on explicitly-undefined behaviour. (Example: ``INCOHERENT`` instance selection.)
+* Does not use ``-Werror`` in its default build configuration.
+* Explicitly specifies a language edition (``Haskell98``, ``GHC2021``), in the source code or the build configuration.
+
```

Comment:
See https://github.com/ghc-proposals/ghc-proposals/pull/625#discussion_r1413559957: stable packages need to have PVP-compliant version bounds, to allow for upgrades to their dependencies that fix compatibility with a newer GHC (e.g. because the dependency itself depends on something experimental).
```suggestion
* Accurately represents its build dependencies according to the PVP.

```

Reply:
What section of the PVP are you referencing here? The spec is very vague:

> When publishing a Cabal package, you SHALL ensure that your dependencies in the build-depends field are accurate. This means specifying not only lower bounds, but also upper bounds on every dependency.

It gives some non-binding guidance later:

> To minimize breakage when new package versions are released, you can use dependencies that are insensitive to minor version changes (e.g. foo >= 1.2.1 && < 1.3). However, note that this approach is slightly risky: when a package exports more things than before, there is a chance that your code will fail to compile due to new name-clash errors. The risk from new name clashes may be small, but you are on the safe side if you import identifiers explicitly or using qualification.

Which indicates that minor release insensitive upper bounds can also cause issues.

It doesn't really specify properly what accurate dependencies are other than "must have any upper bound on all dependencies".

Reply:
This is hard to qualify/articulate. 

A package may very well have lax bounds and never break.

For example, package A depends on package B. If I maintain both of them, I have "total awareness" of their APIs and as such an upper bound may be useless, since I can introduce a revision with an upper bound if I ever break package A.

This is the infamous difference of proactive and reactive upper bounds.

This text **could** demand a proactive upper bound policy: e.g. an upper bound must be at most the latest available package on hackage at the time (e.g. `base < 10000`  would violate that) and successfully build with it.

But that ignores the situation of valid reactive upper bound workflows.

Reply:
My point is that both proactive and reactive are technically valid as per the PVP spec and this has even been a controversial topic amongst hackage trustees: https://github.com/haskell-infra/hackage-trustees/issues/375#issuecomment-1763507820

So my advice is to avoid wording that steers in either direction.

The "ecosystem" part of PVP is very badly written.

I'd go for something like: "A stable package has an upper bound policy that ensures resilience in light of a moving ecosystem and avoids breaking existing build plans.". The latter is a detail that's often overlooked, where revisions can break existing build plans.

You can't machine check this property easily, but I don't think we need to.